staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go

```diff
@@ -1385,7 +1386,7 @@ func TestList(t *testing.T) {
 			expectContinue:             true,
 			expectedRemainingItemCount: utilpointer.Int64Ptr(1),
 			rv:                         "0",
-			expectRV:                   list.ResourceVersion,
+			expectRVFunc:               resourceVersionNoLaterThan(list.ResourceVersion),
```

Comment:
At a minimum my intent was to be able to run the existing tests in parallel on one `etcd` fixture. As that's actually how Kubernetes uses `etcd`, your option 1 is already the case in every Kubernetes cluster that has concurrent access to the database. Tests that fail in such a scenario are IMHO not super useful in documenting the expectations that Kubernetes has on the underlying storage layer.

Reply:
I haven't even read the changed code in this PR (will have to wait until next week), but following the comment thread, I have a couple reactions:

> Tests that fail in such a scenario are IMHO not super useful in documenting the expectations that Kubernetes has on the underlying storage layer.

For a unit test, I disagree. This is bridging two systems, so testing specific expected behavior of the other system in isolation is exactly what I would expect.

> This test assumption does not apply to any non-toy etcd

I would replace "non-toy" with "non-isolated", but testing in isolation is exactly what I want from a unit test.

Reply:
@liggitt why is it useful to document expectations of a _single-threaded_ interaction when no part of the system expects the same or makes any use of the behavior which may occur in that manner? `storage.Interface` and `storage/etcd3.store` are not single-threaded libraries and are not used as such.

Reply:
> @liggitt why is it useful to document expectations of a single-threaded interaction when no part of the system expects the same or makes any use of the behavior which may occur in that manner?

To exercise and verify the etcd behavior the etcd3/store.go implementation is based on (revisions bump on writes, not reads, etc).

> maybe the test could be construed as technically useful for a particular isolated etcd fixture but as there are no (!!) tests for storage.Interface, these tests also IMHO serve as documentation for that interface to the rest of the codebase

if we want to separately provide `storage.Interface` tests (and ensure the etcd3 store implementation passes them), that could be reasonable, but I wouldn't weaken the etcd store unit tests in order to do that.

Reply:
> (revisions bump on writes, not reads, etc).

This is not true today. Unless you specifically ask `etcd` for non-quorum reads, your read bumps the Raft log and the revision.

>  but I wouldn't weaken the etcd store unit tests in order to do that.

Again, this behavior is happenstance for this test fixture and has no implication in any place downstream from etcd. No part of any system - from the `*etcd3.storage` implementation to the end client of the k8s API - ever makes any such assumption.

Reply:
Also I feel like I'm not communicating this appropriately but the `resourceVersion` we give to clients in the `*List` object is the _current global revision for the Raft members_ and therefore changes with _any_ event that requires consensus in the system. The one and only way that this number does not change is if no client interacts with etcd. Absolutely no part of the current `storage/etcd3.store` implementation makes any assumption whatsoever about this behavior, as it's entirely emergent behavior.

Reply:
> This is not true today. When using linearized reads, your request has to go through Raft consensus, if it's in the Raft log the revision is bumped/

that's not accurate. this sets up a 3-member cluster, does linearized reads across them, and demonstrates the revision does not change:

```shell
IP=$(hostname -I | awk '{print $1}')
CLUSTER=etcd0=http://${IP}:2380,etcd1=http://${IP}:2381,etcd2=http://${IP}:2382
for i in $( seq 0 2 ); do
  docker run --rm -d -p 400${i}:400${i} -p 238${i}:238${i} -p 237${i}:237${i} \
    --hostname etcd${i} \
    --name etcd${i} quay.io/coreos/etcd:v3.4.18 \
    etcd \
    -name etcd${i} \
    -advertise-client-urls http://${IP}:237${i},http://${IP}:400${i} \
    -listen-client-urls http://0.0.0.0:237${i},http://0.0.0.0:400${i} \
    -initial-advertise-peer-urls http://${IP}:238${i} \
    -listen-peer-urls http://0.0.0.0:238${i} \
    -initial-cluster-token etcd-cluster-1 \
    -initial-cluster ${CLUSTER} \
    -initial-cluster-state new
done

# write to first member
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2370 etcdctl put /testkey1 testvalue1
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2370 etcdctl put /testkey2 testvalue2
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2370 etcdctl put /testkey3 testvalue3

# fetch from etcd0
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2370 etcdctl get --consistency=l -w json /testkey1
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2370 etcdctl get --consistency=l -w json /testkey2
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2370 etcdctl get --consistency=l -w json /testkey3
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2370 etcdctl get --consistency=l -w json --prefix / | jq .

# fetch from etcd1
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2371 etcdctl get --consistency=l -w json /testkey1
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2371 etcdctl get --consistency=l -w json /testkey2
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2371 etcdctl get --consistency=l -w json /testkey3
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2371 etcdctl get --consistency=l -w json --prefix / | jq .

# fetch from etcd2
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2372 etcdctl get --consistency=l -w json /testkey1
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2372 etcdctl get --consistency=l -w json /testkey2
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2372 etcdctl get --consistency=l -w json /testkey3
ETCDCTL_API=3 ETCDCTL_ENDPOINTS=127.0.0.1:2372 etcdctl get --consistency=l -w json --prefix / | jq .
```

Reply:
> Also I feel like I'm not communicating this appropriately but the resourceVersion we give to clients in the *List object is the current global revision for the Raft members and therefore changes with any event that requires consensus in the system. The one and only way that this number does not change is if no client interacts with etcd. Absolutely no part of the current storage/etcd3.store implementation makes any assumption whatsoever about this behavior, as it's entirely emergent behavior.

Understood. Because the etcd store impl builds things on top of that behavior, that still seems worthwhile to test in isolation.

Reply:
@liggitt hm - my apologies. All of the official documentation says that the linearized read goes through the full Raft process but it looks like they piggy-back on another message so that the read does not technically bump the log. Slightly misleading documentation.

I don't really understand your point, though - our contract for LIST is that we return the current revision, full stop. Testing that "nothing happens" when "nothing is written" is not really the point here.